### PR TITLE
ARROW-13886 [R] Expand documentation for decimal()

### PR DESCRIPTION
**Pull request by dragosmg**
First attempt at this. Will be followed by:
* opening Jira tickets to: 
    * deprecate `decimal()`
    * implement `decimal128()` and `decimal256()`
* expand unit tests both for data types and `Arrays`

Tickets generated:
* implement `decimal128()` - https://issues.apache.org/jira/browse/ARROW-14843
* implement `decimal256()` - https://issues.apache.org/jira/browse/ARROW-14844
* improve messaging around `Decimal128Type` & `Decimal256Type` in the C++ code - https://issues.apache.org/jira/browse/ARROW-14842

### CODE REVIEW

File: r/R/type.R
Comment (thisisnic) [inline r/R/type.R:193]:
```suggestion
#' @param scale For `decimal()`, the number of digits after the decimal
```

File: r/R/type.R
Comment (thisisnic) [inline r/R/type.R:190]:
```suggestion
#' @param precision For `decimal()`, the number of significant digits
```

File: r/R/type.R
Comment (thisisnic) [inline r/R/type.R:371]: I know we use `assert_that()` in a lot of the rest of the code - any reason to not use that here?  Asking only out of curiosity as I think it looks better this way anyway.  Feel free to ignore this picky question!

File: r/R/type.R
Comment (thisisnic) [inline r/R/type.R:367]:
```suggestion
    stop('"precision" must be less than or equal to 38', call. = FALSE)
```

File: r/tests/testthat/test-data-type.R
Comment (thisisnic) [inline r/tests/testthat/test-data-type.R:394]:
```suggestion
  expect_error(decimal(100, 2), "\"precision\" must be less than or equal to 38")
```
nit part 2

File: r/R/type.R
Comment (dragosmg) [reply r/R/type.R]: Done

File: r/R/type.R
Comment (dragosmg) [reply r/R/type.R]: Done

**Review by thisisnic — COMMENTED** [review]
A few pedantic changes but otherwise looks good.  Won't approve yet as I think you're still making changes.

File: r/R/type.R
Comment (dragosmg) [reply r/R/type.R:371]: The function wasn't using `assert_that()` for the other checks and I went with that.

File: r/R/type.R
Comment (dragosmg) [reply r/R/type.R:367]: Done

File: r/tests/testthat/test-data-type.R
Comment (dragosmg) [reply r/tests/testthat/test-data-type.R]: Done

File: r/R/type.R
Comment (thisisnic) [reply r/R/type.R]: Cool!

Comment (dragosmg) [reply r/R/type.R]: Happy to go with `assert_that()` if that's the direction, but I think I should switch all arg check to use it. Maybe we can make a follow-up Jira ticket.

Comment (thisisnic) [reply r/R/type.R]: Nah, it's trivial, I was only asking in case there was a specific thing I didn't know, ultimately it achieves the same thing so don't worry!

File: r/R/type.R
Comment (jonkeane) [inline r/R/type.R:194]: > It can be negative. 

Would it make sense to expand on why one might want to do this? The argument text itself is probably not the right place, but adding a paragraph above in the description might be nice.

File: r/R/type.R
Comment (jonkeane) [inline r/R/type.R:192]: Is decimal256 exposed? Should it be? If it is / if we do, should we mention what the max precision is for it?

File: r/tests/testthat/test-data-type.R
Comment (jonkeane) [inline r/tests/testthat/test-data-type.R:394]: I'm ok adding these, but I'm curious if there's something specifically wrong with the error messages we are seeing from C++ that we're addressing with this re-write of the error message? They seem ~equivalent to me in readability/understandability but I might be too far down this rabbit hole.

_In general_ if the C++ error is understandable, we should defer to that so that we don't need to have a bunch of error handling code duplicated at these two different layers.

Comment (jonkeane) [reply r/tests/testthat/test-data-type.R:394]: For example: when/if we implement other decimal precisions we won't have to gate on each of those in the same way.

**Review by jonkeane — CHANGES_REQUESTED** [review]
Thanks for this, looking good so far, a few questions / comments

File: r/R/type.R
Comment (dragosmg) [reply r/R/type.R:194]: I was debating where to add more info. I agree the arg description isn't the place. Wouldn't it get a bit lost if we put it in the more general `data-type` description?

File: r/R/type.R
Comment (dragosmg) [reply r/R/type.R:192]: decimal256 is not exposed yet. I am planning on opening a bunch of tickets to:
* deprecate `decimal()` and replace it with `decimal128()`
* implement `decimal256()`. 
* maybe open another ticket to agree on a formalised approach to deprecation (both for functions and arguments). It would be relevant both for this PR and for `write_csv_arrow()`

File: r/tests/testthat/test-data-type.R
Comment (dragosmg) [reply r/tests/testthat/test-data-type.R:394]: Agree on consistency. I'm not a huge fan of the current error message (surfaced from C++) as it doesn't tell the user what they need to do to prevent the error. `decimal precision out of range` is not informative if we don't tell them anything about said `range`. It reminds me of frustrating account creating experiences when the password isn't "correct", but you don't know what wrong with it and therefore correction requires a lot of trial and error.

File: r/R/type.R
Comment (dragosmg) [reply r/R/type.R:192]: I added a comment on the ticket - we could record thoughts on implementing `decimal128()` and `decimal256()` there.

File: r/tests/testthat/test-data-type.R
Comment (thisisnic) [reply r/tests/testthat/test-data-type.R:394]: I agree with your point on the phrasing being tighter @dragosmg .  I know some of the C++ folk might prefer us to raise it with them if we can suggest improvements to the error messages that would be good to implement at the C++ level so it benefits everyone - this potentially could be one of those?

File: r/R/type.R
Comment (jonkeane) [inline r/R/type.R:194]: This is really great, thanks! One small suggestion you can take or leave: I think we only need to mention once that the internal encoding is an integer. I like that we include it, but I think it might call more attention than we need having it repeated.

I also really liked the power of 10 analogy / metaphor that you used earlier. It would be nice to include that here (I think it's a concept that arrow users will be familiar with and can grok easily)

File: r/tests/testthat/test-data-type.R
Comment (jonkeane) [reply r/tests/testthat/test-data-type.R:394]: Yeah, we should raise a Jira to suggest a better warning. We don't need to hold this PR to wait for that to happen, but maybe a comment with the Jira # so we know to clean it up if/when the jira happens.

**Review by jonkeane — APPROVED** [review]
Looking good — a few small comments

**Comment by dragosmg** [comment]
Raised a [Jira](https://issues.apache.org/jira/browse/ARROW-14842) to improve the error messaging and reverted to use the error messages as surfaced from the C++ code

File: r/R/type.R
Comment (pitrou) [inline r/R/type.R:183]: Add newline before this?

Comment (dragosmg) [reply r/R/type.R:183]: Sure. I'm actually adding a bit more to that section. Will push shortly

File: r/R/type.R
Comment (dragosmg) [reply r/R/type.R:195]: Done (I think).

File: r/tests/testthat/test-data-type.R
Comment (dragosmg) [reply r/tests/testthat/test-data-type.R]: Done. Restored these to their original state and raised a Jira (first comment) to improve messaging in C++.

Comment (dragosmg) [reply r/tests/testthat/test-data-type.R]: + added a comment with the Jira number in the code

**Pull request closed without merging**
